rust/lance-arrow/src/lib.rs

```diff
@@ -930,12 +931,124 @@ fn merge_list_struct(left: &dyn Array, right: &dyn Array) -> Arc<dyn Array> {
+            if right.null_count() == right.len() {
+                None
+            } else {
+                Some(right.clone())
+            }
+        }
+        (Some(left), Some(right)) => {
+            // Check for placeholders first
+            let left_is_placeholder = left.null_count() == left.len();
+            let right_is_placeholder = right.null_count() == right.len();
+
+            match (left_is_placeholder, right_is_placeholder) {
+                (true, true) => None,                 // Both are placeholders
+                (true, false) => Some(right.clone()), // Left is placeholder, use right
+                (false, true) => Some(left.clone()),  // Right is placeholder, use left
+                (false, false) => {
+                    // Neither is placeholder, perform normal merge
+                    // Fast path: if both have no nulls, can return either one
+                    if left.null_count() == 0 && right.null_count() == 0 {
+                        return Some(left.clone());
+                    }
+
+                    let left_buffer = left.inner();
+                    let right_buffer = right.inner();
+
+                    // Perform bitwise OR directly on BooleanBuffers
+                    // This preserves the correct semantics: 1 = valid, 0 = null
+                    let merged_buffer = left_buffer | right_buffer;
+
+                    Some(arrow_buffer::NullBuffer::from(merged_buffer))
+                }
+            }
+        }
+    }
+}
+
+// Helper function to adjust child array validity based on parent struct validity
+// When parent struct is null, propagates null to child array
+// Optimized with fast paths and SIMD operations
+fn adjust_child_validity(
+    child: &ArrayRef,
+    parent_validity: Option<&arrow_buffer::NullBuffer>,
+) -> ArrayRef {
+    // Fast path: no parent validity means no adjustment needed
+    let parent_validity = match parent_validity {
+        None => return child.clone(),
+        Some(p) if p.null_count() == 0 => return child.clone(), // No nulls to propagate
+        Some(p) => p,
+    };
+
+    let child_validity = child.nulls();
+
+    // Compute the new validity: child_validity AND parent_validity
+    let new_validity = match child_validity {
+        None => {
+            // Fast path: child has no existing validity, just use parent's
+            parent_validity.clone()
+        }
+        Some(child_nulls) => {
+            let child_buffer = child_nulls.inner();
+            let parent_buffer = parent_validity.inner();
+
+            // Perform bitwise AND directly on BooleanBuffers
+            // This preserves the correct semantics: 1 = valid, 0 = null
+            let merged_buffer = child_buffer & parent_buffer;
+
+            arrow_buffer::NullBuffer::from(merged_buffer)
+        }
+    };
+
+    // Create new array with adjusted validity
+    arrow_array::make_array(
+        arrow_data::ArrayData::try_new(
```

Comment:
Hmm, I wonder if we should use `new_unchecked` here if all we are changing is the validity it should be safe?